### ISSUE TITLE
Inflator throws end of compressed block on complete compressed data

### ISSUE THREAD

**Issue by PowerGamer1** [weidai11/cryptopp#112]
Most of my compressed data samples `Inflator` class decompresses just fine. But on some compressed data it throws "Inflator: unexpected end of compressed block" yet the data is valid and complete. Adding one extra byte of ANY VALUE to the end of such data makes `Inflator` to decompress it without errors.

See the attached sample code and data file that reproduces the problem: [default.gz](https://github.com/weidai11/cryptopp/files/87843/default.gz)

P.S. GitHub stuck on uploading sample C++ code in a separate file so I am pasting it here:
```C++
#include <cstdio>
#include "cryptopp/gzip.h"

using namespace CryptoPP;

int main()
{
	byte in[20000], out[70000];
	const size_t file_size = 16782;
	FILE* f = fopen("!.gz", "rb");
	fread(in, 1, file_size, f);
	fclose(f);

	// Works fine:
	//ArraySource(in, file_size, true, new Gunzip(new ArraySink(out, sizeof(out))));

	// GZIP data in test file has 10 bytes of GZIP header and 8 bytes of footer (CRC32 and ISIZE, see http://www.rfc-editor.org/rfc/rfc1952.txt).
	// So we offset "in" by 10 bytes and the size of actual compressed data (without GZIP header and footer) would be (file_size - 18) bytes.
	// Yet the following line of code throws "Inflator: unexpected end of compressed block".
	ArraySource(in + 10, file_size - 18, true, new Inflator(new ArraySink(out, sizeof(out))));

	// If at least one extra byte of data is passed to the Inflator (the value of that byte of data makes no difference) no exception is thrown.
	//in[file_size - 8] = 0xFF; // value of extra byte makes no difference
	//ArraySource(in + 10, file_size - 18 + 1, true, new Inflator(new ArraySink(out, sizeof(out))));
}
```

**Comment by noloader** [comment]
I've always known there's been some minor incompatibilities among inflator programs, but I've never really investigated it. Also see [A lone zero block at ###](http://www-01.ibm.com/support/docview.wss?uid=swg21270445). When using GNU tar, use the `-i` or `--ignore-zeros` option.

OpenSSL experienced a similar issue when using a program called ***`tardy`***; see [Unexpected error when I untar OpenSSL 1.0.2d](https://mta.openssl.org/pipermail/openssl-users/2015-July/001882.html).

**Comment by PowerGamer1** [comment]
@noloader I don't see how the GNU tar and the info in your links relate to what I posted. Tar has nothing at all to do with the sample compressed data in GZIP format I provided. The compressed data is not located at the end of file, it is followed by GZIP footer which is valid and conforms to GZIP spec (http://www.rfc-editor.org/rfc/rfc1952.txt). So its not about some zero bytes at the end of file.

**Comment by noloader** [comment]
@PowerGamer1 - can you provide (1) the source file that's causing trouble, (2) the name of the program that created it, (3) the command used to deflate the source file, and (4) the deflated file?

Sorry to have to ask. I don't really follow what you are doing with the index manipulations in the `ArraySink`. But the comment *"If at least one extra byte of data is passed to the Inflator (the value of that byte of data makes no difference) no exception is thrown"* seems to indicate its a similar problem experienced by GNU tar users.

**Comment by PowerGamer1** [comment]
> @PowerGamer1 - can you provide (1) the source file that's causing trouble, (2) the name of the program that creates the archive, (3) the command used to deflate the source file, and (4) the deflated file?

(1) - you can get the original uncompressed data from the compressed file I attached, using for ex. gzip from http://gzip.org, or by using Crypto++ by uncommenting one of the lines in my sample program that are marked as working without error.

(2), (3) - can't really say what program created archive, the data comes from a server not under my control, but it should not matter anyway - the data is perfectly valid GZIP format data (you can check GZIP header and footer in HEX editor, CryptoPP::Gunzip class accepts them too, and the actual compressed data is valid too, since with workaround with extra byte for Inflator class the data is decompressed and passes GZIP CRC check.).

(4) the sample compressed (deflated) file is attached in my first post (filename is !.gz).

> I don't really follow what you are doing with the index manipulations in the  ArraySink . 

Check the GZIP header and footer (http://www.rfc-editor.org/rfc/rfc1952.txt) of the sample data file I attached (!.gz) and you will understand that I skip them and pass only actual compressed data to Inflator class.

P.S. Is the original author of the library still takes part in its maintenance?

**Comment by noloader** [comment]
> P.S. Is the original author of the library still takes part in its maintenance?

The author of the library is Wei Dai. Unfortunately, no, he does not have time for maintenance and growth of the library. He effectively turned it over to the care of the community.

I'm one of the point-men in the community who has check-in privileges. There are three or four other folks with the same rights. But like with Wei Dai, they don't really have the spare cycles to donate to the project.

As soon as we push Crypto++ 5.7, I'm going back to work.  I took an unpaid unpaid sabbatical to ensure 5.6.3 and 5.7 gets out. I expect 5.7 to be out in the next month or two, so development should slow after that point in time.

**Comment by PowerGamer1** [comment]
The zlib library decompresses the actual compressed data (without GZIP header and footer) from sample file without any problems:
```C++
#include <cstdio>
#include <cassert>

#define ZLIB_CONST
#include "zlib/zlib.h"

int main()
{
	Bytef in[20000], out[70000];
	const size_t file_size = 16782;
	FILE* f = fopen("e:\\!.bin", "rb");
	fread(in, 1, file_size, f);
	fclose(f);

	z_stream stream;
	stream.next_in = in + 10; // skip 10 bytes of GZIP header
	stream.avail_in = file_size - 18; // skip 10 bytes of GZIP header and 8 bytes of GZIP footer
	stream.next_out = out;
	stream.avail_out = sizeof(out);
	stream.zalloc = nullptr;
	stream.zfree = nullptr;

	auto result = inflateInit2(&stream, -MAX_WBITS); // negative windowBits mean no ZLIB or GZIP headers
	assert(result == Z_OK);

	result = inflate(&stream, Z_FINISH);
	assert(result == Z_STREAM_END);

	inflateEnd(&stream);
}
```
So it is definitely a bug in `CryptoPP::Inflator` class.

**Comment by noloader** [comment]
Thanks @PowerGamer1 .

I know Crypto++ attempts to provide a [RFC 1951](http://www.ietf.org/rfc/rfc1951.txt) conforming implementation. I'm trying to locate test vectors for things now.

**Comment by PowerGamer1** [comment]
> test vectors for things

No idea what you are talking about.

**Comment by dirvine** [comment]
Think of test vectors as agreed outputs to fixed inputs to test algorithms.

**Comment by anonimal** [comment]
[We had a similar problem](https://github.com/monero-project/kovri/issues/141) (as did [java I2P](https://github.com/monero-project/kovri/issues/141#issuecomment-190231275)). The issue was resolved by appending a null byte to compressed data. For us, the problem was somehow introduced in Crypto++ 5.6.3 though I can't spot where exactly. Kovri patches [here](https://github.com/anonimal/kovri/commit/8c61f5d2cec8acef8fe7ad0e2a4d172a970681a2) and eventually [here](https://github.com/monero-project/kovri/pull/162/commits/8d63e4dba045d64dad342c32d3e205e500e76944#diff-e502ed0290293c9a661209a0b6e4d6c3R54).

**Comment by noloader** [comment]
Thanks @anonimal. Do you know if the Gzip authors publish any independent test vectors? I have not been able to locate them, and I am hesitant to make a breaking change on others for a special case.

Would it be possible for you to provide a separate `GzipStream` class (following `z_stream`) that handles this issue for you? In this case, it seems like the `NULL` byte should be added automatically when `MessageEnd()` is called on the stream (from your patch):

```
+        decompressor.Put(
+            compressed.data() + '\0',
+            compressed.size() + 1);
```

**Comment by anonimal** [comment]
Hi @noloader. Sorry, but no; I do not know if they have published any independent test vectors. On a related note, Adler himself @madler has a nice response to a [similar question](http://stackoverflow.com/questions/30150972/are-there-test-vectors-available-for-gzip) (though I'm sure you've probably seen this).

@PowerGamer1 I believe the 2nd paragraph (```GzipStream```/```z_stream```) is meant for you (I was confused for a moment :smile:)

Regarding points 1-3: I completely understand the current project's issues.  Now, regarding this particular issue: I've reviewed the diff between tags 5.6.2-5.6.3 for ```zlib/zinflate/cryptlib``` and could not find anything of consequence. Then again, the vast amounts of windows carriage return on the 5.6.3 commits made it difficult to compare actual code change with precision so I may have missed something - or the problem lies elsewhere.

**Comment by noloader** [comment]
> Then again, the vast amounts of windows carriage return on the 5.6.3 commits

@anonimal - Oh that's right... I forgot about that. Let me see if I can do some of the legwork. And I was not aware 5.6.2 did not suffer it.

If we (I) introduced the new behavior, then its something we definitely want to fix. That code had an uninitialized read under Valgrind, and it took me some time to track down. Eventually we pin-pointed it with Coverity.

Since I can't say if I broke it, that points to a gap in out engineering process. We should always be able to detect these things before and after. Guessing after the fact is no good for anyone.

I think I need to fetch Adler's reference implementation, and build out some test cases.

**Comment by noloader** [comment]
@anonimal - Below is a test script that should help with the changes. Things are much more manaeable. Nothing is jumping out at me.

Does anything jump out at you?

-----

```
mkdir cryptopp-gzip
cd cryptopp-gzip
wget https://www.cryptopp.com/cryptopp562.zip
wget https://www.cryptopp.com/cryptopp563.zip
unzip -a cryptopp562.zip -d '5.6.2'
unzip -a cryptopp563.zip -d '5.6.3'

diff --ignore-space-change --ignore-trailing-space 5.6.2/zlib.h 5.6.3/zlib.h
diff --ignore-space-change --ignore-trailing-space 5.6.2/zlib.cpp 5.6.3/zlib.cpp
diff --ignore-space-change --ignore-trailing-space 5.6.2/gzip.h 5.6.3/gzip.h
diff --ignore-space-change --ignore-trailing-space 5.6.2/gzip.cpp 5.6.3/gzip.cpp
```

**Comment by noloader** [comment]
Wait a minute here...

```
**** zinflate.cpp **** 
...
115c120
< 	else if (m_maxCodeBits != 1 && code < (1 << m_maxCodeBits) - blCount[m_maxCodeBits])
---
> 	else if (m_maxCodeBits != 1 && code < shiftedMaxCode - blCount[m_maxCodeBits])
```

Can someone test that?

**Comment by noloader** [comment]
Good news (for PowerGamer). The issue appears to be intermittent, but I can duplicate the issue:

```
$ ./test.exe 
Data: 9480BB7600DDF8A11830D877FB4596A6
Compressed Data: 9BD2B0BB8CE1EE8F85120637CA7FBB4E5B0600
Decompressed stream
$ ./test.exe 
Data: D7229EACCE1B1C0D2F5F8CACA9397C29
Compressed Data: BBAE346FCD3969195EFDF89E352B2D6B3401
Inflator: unexpected end of compressed block
$ ./test.exe 
Data: 58CA88F99248B20ABACC15B3B78BAB71
Compressed Data: 8B38D5F17392C726AE5D6744376FEF5E5D0800
Decompressed stream
$ ./test.exe 
Data: 4C328358FB92C77154D688CCD91FC4F1
Compressed Data: F3316A8EF83DE97861C8B58E3337E58F7C0400
Decompressed stream
$ ./test.exe 
Data: 97A34EF4C3CCD2D4BD037322B1D5E77B
Compressed Data: 9BBED8EFCBE13397AEEC652E56DA78F5793500
Decompressed stream
$ ./test.exe 
Data: B7CA080BB7E28456EE4A87FBB539C9A0
Compressed Data: DB7E8A837BFBA396B0775EEDBFB75A9E5C0000
Decompressed stream
$ ./test.exe 
Data: 5AFED6E676ECC0E69754373217F766E7
Compressed Data: 8BFA77ED59D99B03CFA687981B897F4F7B0E00
Decompressed stream
$ ./test.exe 
Data: 417A6C9DF1246DC1FE4690BFD34E51F8
Compressed Data: 73ACCA99FB5125F7E03FB709FB2FFB05FE0000
Decompressed stream
$ ./test.exe 
Data: 9BFC23E9A61EAC55104740FBE3329123
Compressed Data: 9BFD47F9E532B935A102EE0EBF1F1B4D540600
Decompressed stream
$ ./test.exe 
Data: 8B8B55D9BA5D7604B133CDD632FDB71D
Compressed Data: EBEE0EBDB92BB68C65A3F1D96B467FB7CB0200
Decompressed stream
$ ./test.exe 
Data: 2C1D9C0CC471DB00EDB41EF9EC308791
Compressed Data: D3919DC373A4F036C3DB2D723FDF18B44F0400
Decompressed stream
$ ./test.exe 
Data: 0F16CA19777F53E0C2DE5363D7646B0F
Compressed Data: E3173B25595E1FFCE0D0BDE0E4EB29D9FC00
Inflator: unexpected end of compressed block
$ ./test.exe 
Data: C65E8BC86F0CF63376F8E0265DFF5C5B
Compressed Data: 3B16D77D229FE79B71D98F076AB1FF63A201
Inflator: unexpected end of compressed block
```

-----

```
// g++ $CXXFLAGS -I . test.cxx ./libcryptopp.a -o test.exe
int main(int argc, char* argv[])
{
    try
    {
        AutoSeededRandomPool prng;
        HexEncoder encoder(new FileSink(cout));
        string s1, s2, s3;

        s1.resize(16);
        OS_GenerateRandomBlock(false, (byte*)&s1[0], s1.size());

        cout << "Data: ";
        StringSource ss1(s1, true, new Redirector(encoder));
        cout << endl;

        StringSource ss2(s1, true, new Deflator(new StringSink(s2)));

        cout << "Compressed Data: ";
        StringSource ss3(s2, true, new Redirector(encoder));
        cout << endl;

        StringSource ss4(s2, true, new Inflator(new StringSink(s3)));

        if(s1 != s3)
        {
            cerr << "Failed to decompress stream" << endl;
            return 1;
        }
    
        cout << "Decompressed stream" << endl;
    } catch(const Exception& ex) {
        cerr << ex.what() << endl;
    }

    return 0;
}
```

**Comment by noloader** [comment]
OK, I copied 5.6.2's `*z*` files into 5.6.3 and the problem was not present with the test program. 

I've ruled out the following files by performing a `git checkout <file>.h <file>.cpp` in pairs. That is, undo the copy of the 5.6.2 header and source pair:

* gzip.h
* gzip.cpp
* zlib.h
* zlib.cpp
* zdeflate.h
* zdeflate.cpp

I know its one of:

* zinflate.h
* zinflate.cpp

I'm trying to narrow it down now.

**Comment by noloader** [comment]
I *think* this failure is due to a failure in `reader.FillBuffer` and the subsequent early out from line 212. I added the return value check.

```
bool HuffmanDecoder::Decode(LowFirstBitReader &reader, value_t &value) const
{
	bool result = reader.FillBuffer(m_maxCodeBits);    // Line 211
	if(!result) return false;                          // Line 212

	unsigned int codeBits = Decode(reader.PeekBuffer(), value);
	if (codeBits > reader.BitsBuffered())
		return false;
	reader.SkipBits(codeBits);
	return true;
}
```

**Comment by noloader** [comment]
Fixed at [Commit 60a68714dc3587e9](http://github.com/weidai11/cryptopp/commit/60a68714dc3587e94fa5f428eff8e8789ea72ad8).

**Comment by anonimal** [comment]
Hi @noloader, I'm just now getting these messages. Great catch!
